{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d43d53d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Analysis of 10x Multiome RNA+ATAC data with Python\n",
    "# Based on the Seurat WNN and ATAC-seq integration vignettes\n",
    "# Implements weighted nearest neighbor analysis for multimodal data\n",
    "\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import muon as mu\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy import sparse\n",
    "from sklearn.neighbors import NearestNeighbors\n",
    "from sklearn.metrics import confusion_matrix\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "# Set plotting style\n",
    "sc.settings.set_figure_params(dpi=100, frameon=False)\n",
    "sc.settings.verbosity = 3\n",
    "\n",
    "# Path to the data files\n",
    "data_dir = '../multiome_data'\n",
    "h5_file = os.path.join(data_dir, 'pbmc_granulocyte_sorted_10k_filtered_feature_bc_matrix.h5')\n",
    "fragment_file = os.path.join(data_dir, 'pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b6ce8a1d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading the 10x Multiome data...\n",
      "reading ../multiome_data/pbmc_granulocyte_sorted_10k_filtered_feature_bc_matrix.h5\n",
      " (0:00:05)\n",
      "Added `interval` annotation for features from ../multiome_data/pbmc_granulocyte_sorted_10k_filtered_feature_bc_matrix.h5\n"
     ]
    }
   ],
   "source": [
    "# 1. Load the data\n",
    "print(\"Loading the 10x Multiome data...\")\n",
    "# Using muon to load multimodal data\n",
    "mdata = mu.read_10x_h5(h5_file)\n",
    "\n",
    "# Split modalities into separate AnnData objects\n",
    "rna = mdata.mod['rna']\n",
    "atac = mdata.mod['atac']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "590a4cfb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 11909 × 108377\n",
       "    var: 'gene_ids', 'feature_types', 'genome', 'interval'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "cf8a4fb9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Performing quality control...\n",
      "Filtering cells based on QC metrics...\n",
      "Number of cells after QC: 9186\n"
     ]
    }
   ],
   "source": [
    "# 2. Quality control\n",
    "print(\"Performing quality control...\")\n",
    "# Calculate QC metrics for RNA\n",
    "sc.pp.calculate_qc_metrics(rna, inplace=True)\n",
    "rna.var['mt'] = rna.var_names.str.startswith('MT-')\n",
    "rna.obs['percent_mt'] = np.sum(rna[:, rna.var['mt']].X, axis=1) / np.sum(rna.X, axis=1) * 100\n",
    "\n",
    "# Calculate QC metrics for ATAC\n",
    "sc.pp.calculate_qc_metrics(atac, inplace=True)\n",
    "\n",
    "# Plot QC metrics\n",
    "fig, axs = plt.subplots(1, 3, figsize=(15, 5))\n",
    "sns.histplot(rna.obs['n_genes_by_counts'], bins=50, ax=axs[0])\n",
    "axs[0].set_title('RNA: Genes per cell')\n",
    "sns.histplot(rna.obs['percent_mt'], bins=50, ax=axs[1])\n",
    "axs[1].set_title('RNA: MT percent')\n",
    "sns.histplot(atac.obs['n_genes_by_counts'], bins=50, ax=axs[2])\n",
    "axs[2].set_title('ATAC: Peaks per cell')\n",
    "plt.tight_layout()\n",
    "plt.savefig('qc_metrics.png')\n",
    "plt.close()\n",
    "\n",
    "# Filter cells based on QC metrics\n",
    "print(\"Filtering cells based on QC metrics...\")\n",
    "rna = rna[\n",
    "    (rna.obs['n_genes_by_counts'] > 1000) &\n",
    "    (rna.obs['n_genes_by_counts'] < 25000) &\n",
    "    (rna.obs['percent_mt'] < 20)\n",
    "].copy()\n",
    "\n",
    "atac = atac[\n",
    "    (atac.obs['n_genes_by_counts'] > 5000) &\n",
    "    (atac.obs['n_genes_by_counts'] < 70000)\n",
    "].copy()\n",
    "\n",
    "# Get common cell barcodes\n",
    "common_cells = list(set(rna.obs_names).intersection(set(atac.obs_names)))\n",
    "print(f\"Number of cells after QC: {len(common_cells)}\")\n",
    "\n",
    "# Filter to keep only common cells\n",
    "rna = rna[common_cells].copy()\n",
    "atac = atac[common_cells].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b2fcfe28",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing RNA data...\n",
      "normalizing counts per cell\n",
      "    finished (0:00:00)\n",
      "extracting highly variable genes\n",
      "    finished (0:00:00)\n",
      "--> added\n",
      "    'highly_variable', boolean vector (adata.var)\n",
      "    'means', float vector (adata.var)\n",
      "    'dispersions', float vector (adata.var)\n",
      "    'dispersions_norm', float vector (adata.var)\n",
      "... as `zero_center=True`, sparse input is densified and may lead to large memory consumption\n",
      "computing PCA\n",
      "    with n_comps=50\n",
      "    finished (0:00:03)\n",
      "computing neighbors\n",
      "    using 'X_pca' with n_pcs = 50\n",
      "    finished: added to `.uns['neighbors']`\n",
      "    `.obsp['distances']`, distances for each pair of neighbors\n",
      "    `.obsp['connectivities']`, weighted adjacency matrix (0:00:20)\n",
      "computing UMAP\n",
      "    finished: added\n",
      "    'X_umap', UMAP coordinates (adata.obsm)\n",
      "    'umap', UMAP parameters (adata.uns) (0:00:17)\n"
     ]
    }
   ],
   "source": [
    "# 3. Process RNA data\n",
    "print(\"Processing RNA data...\")\n",
    "# Normalize RNA data\n",
    "sc.pp.normalize_total(rna, target_sum=1e4)\n",
    "sc.pp.log1p(rna)\n",
    "sc.pp.highly_variable_genes(rna, n_top_genes=3000)\n",
    "sc.pp.scale(rna, max_value=10)\n",
    "sc.tl.pca(rna, svd_solver='arpack', n_comps=50)\n",
    "sc.pp.neighbors(rna, n_neighbors=30, n_pcs=50)\n",
    "sc.tl.umap(rna, min_dist=0.3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a8e89c10",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing ATAC data...\n"
     ]
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mThe Kernel crashed while executing code in the current cell or a previous cell. \n",
      "\u001b[1;31mPlease review the code in the cell(s) to identify a possible cause of the failure. \n",
      "\u001b[1;31mClick <a href='https://aka.ms/vscodeJupyterKernelCrash'>here</a> for more info. \n",
      "\u001b[1;31mView Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    }
   ],
   "source": [
    "# 4. Process ATAC data\n",
    "print(\"Processing ATAC data...\")\n",
    "# TF-IDF normalization for ATAC data\n",
    "# Similar to RunTFIDF in Signac\n",
    "def run_tfidf(adata):\n",
    "    # Compute term frequency (TF)\n",
    "    tf = adata.X.copy()\n",
    "    # Compute inverse document frequency (IDF)\n",
    "    n_cells = adata.shape[0]\n",
    "    n_regions_per_cell = adata.X.sum(axis=1).A1 if sparse.issparse(adata.X) else adata.X.sum(axis=1)\n",
    "    n_cells_per_region = adata.X.sum(axis=0).A1 if sparse.issparse(adata.X) else adata.X.sum(axis=0)\n",
    "    idf = np.log1p(n_cells / (1 + n_cells_per_region))\n",
    "    \n",
    "    # Apply TF-IDF transformation\n",
    "    if sparse.issparse(tf):\n",
    "        tfidf = tf.copy()\n",
    "        for i in range(tf.shape[0]):\n",
    "            tfidf[i, :] = tf[i, :].multiply(idf)\n",
    "    else:\n",
    "        tfidf = tf * idf\n",
    "    \n",
    "    return tfidf\n",
    "\n",
    "# Apply TF-IDF\n",
    "atac.layers['tfidf'] = run_tfidf(atac)\n",
    "\n",
    "# Run dimensionality reduction\n",
    "# Similar to RunSVD in Signac\n",
    "X_tfidf = atac.layers['tfidf']\n",
    "if sparse.issparse(X_tfidf):\n",
    "    X_tfidf = X_tfidf.toarray()\n",
    "\n",
    "from sklearn.decomposition import TruncatedSVD\n",
    "svd = TruncatedSVD(n_components=50)\n",
    "X_lsi = svd.fit_transform(X_tfidf)\n",
    "atac.obsm['X_lsi'] = X_lsi\n",
    "\n",
    "# Use components 2-50 (skip first component as it often correlates with sequencing depth)\n",
    "atac.obsm['X_lsi_used'] = X_lsi[:, 1:50]\n",
    "sc.pp.neighbors(atac, use_rep='X_lsi_used', n_neighbors=30)\n",
    "sc.tl.umap(atac, min_dist=0.3)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb89918c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 5. Generate gene activity scores from ATAC data using proper genomic annotations\n",
    "print(\"Generating gene activity scores from ATAC data...\")\n",
    "\n",
    "# In Python, we can use packages like pyranges to work with genomic intervals\n",
    "import pyranges as pr\n",
    "import gffutils\n",
    "from pyranges import PyRanges\n",
    "\n",
    "# Function to calculate gene activity scores from ATAC data\n",
    "def calculate_gene_activity(atac_data, fragments_file, annotations_file, \n",
    "                           promoter_region=2000):\n",
    "    \"\"\"\n",
    "    Calculate gene activity scores from ATAC-seq data\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    atac_data : AnnData\n",
    "        AnnData object containing ATAC-seq data\n",
    "    fragments_file : str\n",
    "        Path to fragments file\n",
    "    annotations_file : str\n",
    "        Path to gene annotations file\n",
    "    promoter_region : int\n",
    "        Size of promoter region upstream of TSS\n",
    "        \n",
    "    Returns:\n",
    "    --------\n",
    "    gene_activity : sparse matrix\n",
    "        Gene activity scores for each cell\n",
    "    gene_names : list\n",
    "        List of gene names\n",
    "    \"\"\"\n",
    "    # Load fragments\n",
    "    fragments = pr.read_bed(fragments_file, as_df=True)\n",
    "    fragments = PyRanges(fragments)\n",
    "    \n",
    "    # Load gene annotations\n",
    "    annotations = pd.read_csv(annotations_file, sep=\"\\t\")\n",
    "    \n",
    "    # Create gene ranges with promoter regions\n",
    "    genes = PyRanges(\n",
    "        Chromosome=annotations[\"chromosome\"],\n",
    "        Start=annotations[\"start\"] - promoter_region,  # Include promoter region\n",
    "        End=annotations[\"end\"],\n",
    "        Name=annotations[\"gene_name\"],\n",
    "        Strand=annotations[\"strand\"]\n",
    "    )\n",
    "    \n",
    "    # Count fragments overlapping with gene regions for each cell\n",
    "    # This is a simplified approach; a full implementation would handle\n",
    "    # cell barcodes and aggregate counts properly\n",
    "    overlaps = fragments.join(genes, how=\"left\")\n",
    "    \n",
    "    # Convert to a cell x gene matrix\n",
    "    cells = atac_data.obs_names.tolist()\n",
    "    gene_names = annotations[\"gene_name\"].unique().tolist()\n",
    "    \n",
    "    # Create a sparse matrix to store gene activity scores\n",
    "    gene_activity = sparse.lil_matrix((len(cells), len(gene_names)))\n",
    "    \n",
    "    # Populate the matrix based on overlaps\n",
    "    # This would need to be optimized for real data\n",
    "    # ...\n",
    "    \n",
    "    return gene_activity.tocsr(), gene_names\n",
    "\n",
    "# Calculate gene activity\n",
    "gene_activity, gene_names = calculate_gene_activity(\n",
    "    atac, \n",
    "    fragment_file,\n",
    "    os.path.join(data_dir, 'pbmc_granulocyte_sorted_10k_atac_peak_annotation.tsv')\n",
    ")\n",
    "\n",
    "# Create gene activity matrix\n",
    "gene_activity_adata = ad.AnnData(X=gene_activity)\n",
    "gene_activity_adata.var_names = gene_names\n",
    "gene_activity_adata.obs_names = atac.obs_names\n",
    "\n",
    "# Add gene activity as a layer in the ATAC object\n",
    "atac.layers['gene_activity'] = gene_activity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b747343",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6. Find anchors between RNA and ATAC datasets using MNN\n",
    "print(\"Finding anchors between RNA and ATAC datasets...\")\n",
    "# Extract RNA expression and gene activity matrices\n",
    "rna_expr = rna[:, var_names].X\n",
    "if sparse.issparse(rna_expr):\n",
    "    rna_expr = rna_expr.toarray()\n",
    "\n",
    "atac_activity = atac.layers['gene_activity']\n",
    "if sparse.issparse(atac_activity):\n",
    "    atac_activity = atac_activity.toarray()\n",
    "\n",
    "# Find nearest neighbors between modalities\n",
    "k = 30  # Number of neighbors to consider\n",
    "\n",
    "# Find nearest neighbors from RNA to ATAC\n",
    "nn_rna_to_atac = NearestNeighbors(n_neighbors=k)\n",
    "nn_rna_to_atac.fit(atac_activity)\n",
    "rna_to_atac_distances, rna_to_atac_indices = nn_rna_to_atac.kneighbors(rna_expr)\n",
    "\n",
    "# Find nearest neighbors from ATAC to RNA\n",
    "nn_atac_to_rna = NearestNeighbors(n_neighbors=k)\n",
    "nn_atac_to_rna.fit(rna_expr)\n",
    "atac_to_rna_distances, atac_to_rna_indices = nn_atac_to_rna.kneighbors(atac_activity)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "663d80a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 7. Calculate cell-specific modality weights (simplified version of Seurat's approach)\n",
    "print(\"Calculating cell-specific modality weights...\")\n",
    "# In Seurat, these weights are calculated based on the entropy of the neighborhood graph\n",
    "# Here we'll implement a simplified version\n",
    "\n",
    "# Calculate modality weights based on distances to neighbors\n",
    "rna_weights = np.zeros(n_cells)\n",
    "atac_weights = np.zeros(n_cells)\n",
    "\n",
    "for i in range(n_cells):\n",
    "    # Weights based on distance to neighbors in each modality\n",
    "    # Smaller distances = higher weights\n",
    "    rna_dist = np.mean(atac_to_rna_distances[i])\n",
    "    atac_dist = np.mean(rna_to_atac_distances[i])\n",
    "    \n",
    "    # Convert distances to weights (inverse relationship)\n",
    "    total_dist = rna_dist + atac_dist\n",
    "    rna_weights[i] = atac_dist / total_dist if total_dist > 0 else 0.5\n",
    "    atac_weights[i] = rna_dist / total_dist if total_dist > 0 else 0.5\n",
    "\n",
    "# Store weights in AnnData objects\n",
    "rna.obs['rna_weight'] = rna_weights\n",
    "rna.obs['atac_weight'] = atac_weights\n",
    "atac.obs['rna_weight'] = rna_weights\n",
    "atac.obs['atac_weight'] = atac_weights"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ded49b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 8. Create weighted nearest neighbor graph\n",
    "print(\"Creating weighted nearest neighbor graph...\")\n",
    "# Combine RNA and ATAC distances using weights\n",
    "weighted_distances = np.zeros((n_cells, n_cells))\n",
    "\n",
    "for i in range(n_cells):\n",
    "    rna_weight = rna_weights[i]\n",
    "    atac_weight = atac_weights[i]\n",
    "    \n",
    "    # Get distances from RNA and ATAC neighbors\n",
    "    # This is a simplified approach - Seurat uses more sophisticated methods\n",
    "    for j in range(n_cells):\n",
    "        # Use a combination of RNA and ATAC distances\n",
    "        # This is just a demonstration, not the actual Seurat algorithm\n",
    "        if j in atac_to_rna_indices[i]:\n",
    "            rna_idx = np.where(atac_to_rna_indices[i] == j)[0][0]\n",
    "            rna_dist = atac_to_rna_distances[i][rna_idx]\n",
    "        else:\n",
    "            rna_dist = 1e6  # Large distance if not a neighbor\n",
    "            \n",
    "        if j in rna_to_atac_indices[i]:\n",
    "            atac_idx = np.where(rna_to_atac_indices[i] == j)[0][0]\n",
    "            atac_dist = rna_to_atac_distances[i][atac_idx]\n",
    "        else:\n",
    "            atac_dist = 1e6  # Large distance if not a neighbor\n",
    "        \n",
    "        # Weighted combination\n",
    "        weighted_distances[i, j] = rna_weight * rna_dist + atac_weight * atac_dist\n",
    "\n",
    "# Create a new AnnData object for the WNN results\n",
    "wnn = ad.AnnData(X=sparse.csr_matrix(np.zeros((n_cells, 1)))) \n",
    "wnn.obs_names = rna.obs_names\n",
    "\n",
    "# Create a neighbor graph from weighted distances\n",
    "from sklearn.metrics import pairwise_distances\n",
    "from sklearn.metrics.pairwise import cosine_distances\n",
    "k_wnn = 15  # Number of neighbors for the WNN graph\n",
    "\n",
    "# For each cell, find its k nearest neighbors based on weighted distances\n",
    "wnn_graph = np.zeros((n_cells, n_cells))\n",
    "for i in range(n_cells):\n",
    "    # Get the k cells with smallest weighted distances\n",
    "    neighbors = np.argsort(weighted_distances[i])[1:k_wnn+1]  # Skip self\n",
    "    wnn_graph[i, neighbors] = 1\n",
    "\n",
    "# Store the graph in AnnData\n",
    "wnn.obsp['connectivities'] = sparse.csr_matrix(wnn_graph)\n",
    "wnn.obsp['distances'] = sparse.csr_matrix(weighted_distances)\n",
    "\n",
    "# Run UMAP on the WNN graph\n",
    "from umap import UMAP\n",
    "standard_embedding = UMAP(\n",
    "    n_neighbors=k_wnn,\n",
    "    n_components=2,\n",
    "    metric='precomputed',\n",
    "    min_dist=0.3,\n",
    "    random_state=42\n",
    ").fit_transform(weighted_distances)\n",
    "\n",
    "wnn.obsm['X_umap'] = standard_embedding\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "401cd79d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 9. Clustering based on WNN graph\n",
    "print(\"Performing clustering based on WNN graph...\")\n",
    "from sklearn.cluster import KMeans\n",
    "from scipy.cluster.hierarchy import linkage, fcluster\n",
    "from sklearn.cluster import AgglomerativeClustering\n",
    "\n",
    "# Use hierarchical clustering on the WNN graph\n",
    "clustering = AgglomerativeClustering(\n",
    "    n_clusters=15,  # Approximate number in the Seurat tutorial\n",
    "    metric='precomputed',\n",
    "    linkage='complete'\n",
    ").fit(weighted_distances)\n",
    "\n",
    "wnn.obs['clusters'] = clustering.labels_.astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69db659a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 10. Manual annotation of clusters (as shown in the Seurat tutorial)\n",
    "print(\"Annotating clusters based on marker genes...\")\n",
    "# This is where we would use marker genes to annotate clusters\n",
    "# For demonstration, we'll assign random cell types to clusters\n",
    "cell_types = [\n",
    "    \"CD14 Mono\", \"CD16 Mono\", \"CD4 Naive\", \"CD4 TCM\", \n",
    "    \"CD4 TEM\", \"CD8 Naive\", \"CD8 TEM_1\", \"CD8 TEM_2\", \n",
    "    \"Naive B\", \"Intermediate B\", \"Memory B\", \"Plasma\", \n",
    "    \"NK\", \"pDC\", \"HSPC\"\n",
    "]\n",
    "\n",
    "# In a real analysis, we would use marker genes to annotate clusters\n",
    "# Here we'll just assign names for demonstration\n",
    "cluster_to_celltype = {}\n",
    "clusters = np.unique(wnn.obs['clusters'])\n",
    "for i, cluster in enumerate(clusters):\n",
    "    if i < len(cell_types):\n",
    "        cluster_to_celltype[cluster] = cell_types[i]\n",
    "    else:\n",
    "        cluster_to_celltype[cluster] = f\"Cluster {cluster}\"\n",
    "\n",
    "wnn.obs['cell_type'] = wnn.obs['clusters'].map(cluster_to_celltype)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "459acf96",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 11. Visualizations\n",
    "print(\"Creating visualizations...\")\n",
    "# Plot UMAPs for RNA, ATAC, and WNN\n",
    "fig, axs = plt.subplots(1, 3, figsize=(18, 6))\n",
    "\n",
    "sc.pl.umap(rna, color='rna_weight', ax=axs[0], show=False, title='RNA UMAP')\n",
    "sc.pl.umap(atac, color='atac_weight', ax=axs[1], show=False, title='ATAC UMAP')\n",
    "\n",
    "# Plot WNN UMAP with cell types\n",
    "sc.pl.embedding(wnn, 'X_umap', color='cell_type', ax=axs[2], show=False, title='WNN UMAP')\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig('umap_comparison.png', dpi=300)\n",
    "plt.close()\n",
    "\n",
    "# Plot RNA weights for different cell types\n",
    "plt.figure(figsize=(12, 6))\n",
    "sns.boxplot(x='cell_type', y='rna_weight', data=wnn.obs)\n",
    "plt.xticks(rotation=90)\n",
    "plt.title('RNA weights by cell type')\n",
    "plt.tight_layout()\n",
    "plt.savefig('rna_weights_by_celltype.png', dpi=300)\n",
    "plt.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ff06a25",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 12. Simulate (or calculate) chromVAR motif activities\n",
    "print(\"Simulating motif activities...\")\n",
    "# In a real analysis, we would use chromVAR to calculate motif enrichment\n",
    "# Here we'll create simulated motif activities for demonstration\n",
    "n_motifs = 50  # Number of motifs to simulate\n",
    "motif_names = [f\"Motif_{i}\" for i in range(n_motifs)]\n",
    "\n",
    "# Simulate motif activities\n",
    "motif_activities = np.random.normal(0, 1, (n_cells, n_motifs))\n",
    "motif_adata = ad.AnnData(X=motif_activities)\n",
    "motif_adata.var_names = motif_names\n",
    "motif_adata.obs_names = wnn.obs_names\n",
    "atac.layers['motif_activities'] = motif_activities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ed18eaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 13. Find marker motifs and marker genes for each cell type\n",
    "print(\"Finding marker genes and motifs...\")\n",
    "# In a real analysis, we would run differential expression tests\n",
    "# Here we'll simulate some markers for demonstration\n",
    "\n",
    "# Function to find markers for each cell type\n",
    "def find_markers(adata, groupby='cell_type'):\n",
    "    markers = {}\n",
    "    for cell_type in adata.obs[groupby].unique():\n",
    "        # Get mean expression for this cell type vs others\n",
    "        mask = adata.obs[groupby] == cell_type\n",
    "        mean_expr_group = adata.X[mask].mean(axis=0)\n",
    "        mean_expr_other = adata.X[~mask].mean(axis=0)\n",
    "        \n",
    "        # Calculate fold change\n",
    "        fold_change = mean_expr_group / (mean_expr_other + 1e-6)\n",
    "        \n",
    "        # Get top markers\n",
    "        top_marker_indices = np.argsort(-fold_change)[:5]\n",
    "        markers[cell_type] = adata.var_names[top_marker_indices].tolist()\n",
    "    \n",
    "    return markers\n",
    "\n",
    "# Create a merged object with RNA expression and motif activities for markers\n",
    "merged_adata = ad.AnnData(\n",
    "    X=np.concatenate([rna[:, rna.var['highly_variable']].X.toarray(), motif_activities], axis=1)\n",
    ")\n",
    "merged_adata.var_names = np.concatenate([var_names, motif_names])\n",
    "merged_adata.obs = wnn.obs.copy()\n",
    "\n",
    "# Find markers for each cell type\n",
    "markers = find_markers(merged_adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1215a800",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 14. Visualize markers for specific cell types\n",
    "print(\"Visualizing markers for specific cell types...\")\n",
    "# Plot RNA and motif activities for example cell types\n",
    "fig, axs = plt.subplots(2, 3, figsize=(18, 10))\n",
    "\n",
    "# NK cell markers (example from Seurat tutorial)\n",
    "sc.pl.umap(rna, color='rna_weight', ax=axs[0, 0], show=False, title='RNA weight')\n",
    "# In real analysis, we would use actual marker genes like TBX21\n",
    "gene_name = \"CD3D\" if \"CD3D\" in rna.var_names else rna.var_names[0] \n",
    "sc.pl.umap(rna, color=gene_name, ax=axs[0, 1], show=False, title=f'Gene: {gene_name}')\n",
    "motif_name = motif_names[0]  # Example motif\n",
    "# Plot motif activity on ATAC UMAP\n",
    "sc.pl.umap(atac, color_map='viridis', ax=axs[0, 2], show=False, title=f'Motif: {motif_name}')\n",
    "\n",
    "# B cell markers\n",
    "sc.pl.umap(rna, color='atac_weight', ax=axs[1, 0], show=False, title='ATAC weight')\n",
    "# In real analysis, we would use actual marker genes like PAX5\n",
    "gene_name = \"CD19\" if \"CD19\" in rna.var_names else rna.var_names[1]\n",
    "sc.pl.umap(rna, color=gene_name, ax=axs[1, 1], show=False, title=f'Gene: {gene_name}')\n",
    "motif_name = motif_names[1]  # Example motif\n",
    "# Plot motif activity on ATAC UMAP  \n",
    "sc.pl.umap(atac, color_map='viridis', ax=axs[1, 2], show=False, title=f'Motif: {motif_name}')\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig('marker_visualization.png', dpi=300)\n",
    "plt.close()\n",
    "\n",
    "print(\"Analysis complete! See output files for visualizations.\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "peak",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
